README.rst

pytv for Python
===================

This library uses the `tvmaze <http://www.tvmaze.com/api>`_
API. All data is credited to them.

This
documentation
is a work
in progress


Installation:

    pip install pytv

Usage

    from pytv import Show, Schedule


    Get schedule for current day::

        >>>schedule = Schedule()
        >>>schedule.episodes
        [list of episodes]

    Get schedule for other days::

        >>>schedule = Schedule(date='2011-25-12')

    Get a show::

        >>>show = Show(show_id=1)
        >>>show.name
        'Under The Dome'

The Show class has the following attributes:

- episodes
- specials
- seasons
- cast
- crew
- next_episode
- summary
- name
- premiered

Search class:

    init: variant, query, lookup

    query: required if variant is in [show, singlesearch, people]

    lookup: required if variant is lookup

    variant options:

    - show(default)
    - singlesearch
    - lookup
    - people

    attributes:

    - results (list if variant in [show, people] single object if other)

Search usage

  from pytv import Search

  search shows::

    >>>search = Search(query='breaking bad')
    >>>search.results
    [list of results as Show objects]

  lookup tvdb id::

    >>>search = Search(variant="lookup", lookup="thetvdb=81189")
    >>>search.results
    (Show object)



CHANGELOG.md

# Pytv ChangeLog
___

## v0.3.0 (2018-03-10)

#### Fixed:
* proper error on 404

### Added:
* Search class
* Shows class
* get_updates

## v0.2.1 (2017-10-03)

#### Fixed:

* Show.cast properly returns list of cast members as Cast objects when initialized Show with embed_url with cast
* Show.crew properly returns list oc crew members as Crew objects when initialized Show with embed_url with crew

## v0.2.0 (2017-09-16)

### New Classes:

* Episode
* Cast
* Crew

### Added:
* Schedule.episodes and Show.episodes now return a list of episodes as Episode objects
* Show.next_episode returns Episode object of the next episode to air
* Show.get_episode(int_here) returns Episode object based on number passed in or none if out of range


#### Fixed:

* Schedule().episodes properly errors if it's a bad date


pytv/tvmaze_tests.py
"""tests for tvmaze.py"""

import requests
import time
from pytv.tvmaze import (
    Schedule,
    Shows,
    Show,
    Search,
    Season,
    Cast,
    Crew,
    Person,
    Episode,
    ApiError,
    make_request,
    get_updates
)
from .apiurls import SHOW, SHOWS
import unittest


class TestSchedule(unittest.TestCase):
    """Tests Schedule class in tvmaze.py"""

    def setUp(self):
        """Creates a schedule"""
        self.schedule = Schedule()

    def test_init_schedule(self):
        """Tests new Schedule class"""

        # test country code default is set to us
        self.assertEqual('US', self.schedule.country_code)
        self.assertTrue(self.schedule.date, time.strftime('%Y-%m-%d'))
        self.assertTrue(self.schedule.url)

    def test_schedule_has_episodes(self):
        """Tests Schedule gets episodes from the tvmaze api"""
        response = requests.get(self.schedule.url)

        # test api returns 200
        self.assertEqual(200, response.status_code)
        episodes = self.schedule.episodes

        self.assertTrue(episodes)

    def test_include_networks(self):
        """include_networks() should return all episodes in the included networks"""
        new_episodes = self.schedule.include_networks(['Disney Channel'])
        for episode in new_episodes:
            self.assertTrue(episode.show.network["name"] in ['Disney Channel'])

    def test_include_multiple_networks(self):
        """include_networks() should return all episodes with multiple networks included"""
        networks = ['Disney Channel', 'HGTV', 'CBS']
        new_episodes = self.schedule.include_networks(networks)
        for episode in new_episodes:
            self.assertTrue(episode.show.network["name"] in networks)


class TestShows(unittest.TestCase):

    def get_compare_data(self, page=0):
        return [Show(**show) for show in make_request('{}?page={}'.format(SHOWS, page))]

    def compare_api_and_class(self, show_class):
        shows_data = self.get_compare_data(show_class.page)
        for show1, show2 in zip(show_class.shows, shows_data):
            self.assertEqual(show1.id, show2.id)
            self.assertEqual(show1.name, show2.name)

    def test_shows(self):
        shows = Shows()
        self.compare_api_and_class(shows)

    def test_shows_with_page(self):
        shows = Shows(page=3)
        self.assertEqual(shows.page, 3)
        self.compare_api_and_class(shows)

    def test_shows_next_page(self):
        shows = Shows(page=6)
        shows.next_page()
        self.assertEqual(shows.page, 7)
        self.compare_api_and_class(shows)

    def test_shows_previous_page(self):
        shows = Shows(page=5)
        shows.previous_page()
        self.assertEqual(shows.page, 4)
        self.compare_api_and_class(shows)

    def test_repeat_pagination_calls(self):
        shows = Shows(page=3)
        shows.next_page()
        shows.next_page()
        self.assertEqual(shows.page, 5)
        self.compare_api_and_class(shows)
        shows.next_page()
        shows.next_page()
        self.assertEqual(shows.page, 7)
        self.compare_api_and_class(shows)
        shows.previous_page()
        self.assertEqual(shows.page, 6)
        self.compare_api_and_class(shows)
        shows.previous_page()
        shows.previous_page()
        self.assertEqual(shows.page, 4)
        self.compare_api_and_class(shows)
        shows.previous_page()
        self.assertEqual(shows.page, 3)
        self.compare_api_and_class(shows)

    def test_out_of_range_page(self):
        """A page below 0 should return page 0 a page above max pages should raise error"""
        shows = Shows(page=-1)
        self.assertEqual(0, shows.page)
        self.compare_api_and_class(shows)

        self.assertRaises(ApiError, lambda: Shows(page=1000))

        shows = Shows(page=141)
        self.assertRaises(ApiError, lambda: shows.next_page())
        self.assertEqual(shows.page, 141)

    def test_updates(self):
        data = get_updates()
        self.assertTrue(data)
        for key in data.keys():
            self.assertTrue(int(key))


class TestShow(unittest.TestCase):
    """Tests Show class in tvmaze.py"""

    def test_create_show(self):
        """Tests init method in show class"""
        show = Show(show_id=1)
        self.assertEqual('http://api.tvmaze.com/shows/1', show.api_url)
        self.assertEqual(1, show.id)

    def test_bad_create_show(self):
        """init method of show class should raise error if show id is not valid"""
        self.assertRaises(ApiError, lambda: Show(show_id='3fad'))

    def test_create_show_from_api_data(self):
        data = make_request('{}1'.format(SHOW))
        show = Show(**data)
        self.assertEqual(data['id'], show.id)
        self.assertEqual(data['id'], show.show_id)
        self.assertDictEqual(data['schedule'], show.schedule)
        self.assertDictEqual(data['network'], show.network)
        self.assertListEqual(data['genres'], show.genres)

    def test_create_show_with_embed_cast_url(self):
        """Tests init method in show class with additional arg embed_url"""
        show = Show(show_id=1, embed_url='?embed=cast')
        self.assertEqual('http://api.tvmaze.com/shows/1?embed=cast', show.api_url)

    def test_create_show_with_embed_season_url(self):
        """Init method in show class when embed_url=seasons should return all seasons as Season class per season"""
        show = Show(show_id=1, embed_url='?embed=seasons')
        self.assertIsInstance(show.seasons[0], Season)
        self.assertTrue(show.seasons[0].episodes)
        self.assertEqual(1, show.seasons[0].number)

    def test_episodes(self):
        """episodes property should return a list of episodes"""
        show = Show(show_id=1)
        episodes = show.episodes
        for episode in episodes:
            self.assertTrue('under-the-dome' in episode.url)

    def test_seasons(self):
        """seasons property should return list of seasons"""
        show = Show(show_id=1)
        self.assertFalse(show.season_list)
        seasons = show.seasons
        self.assertTrue(show.season_list)

        for season in seasons:
            self.assertIsInstance(season, Season)

    def test_specials(self):
        """specials property should return list of specials as Episode objects"""
        show = Show(show_id=1)
        specials = show.specials
        self.assertTrue(specials)
        for episode in specials:
            self.assertIsInstance(episode, Episode)

    def test_failed_episode_by_season_and_number(self):
        """episode_by_number() should raise ValueError if bad season or episode"""
        show = Show(show_id=1)
        self.assertRaises(ValueError, lambda: show.episode_by_number(5, 1))
        self.assertRaises(ValueError, lambda: show.episode_by_number(1, 25))

    def test_episode_by_season_and_number(self):
        """episode_by_number() should return dict of episode information"""
        show = Show(show_id=1)
        episode = show.episode_by_number(1, 1)
        self.assertEqual(1, episode['number'])
        self.assertEqual(1, episode['season'])

        second_episode = show.episode_by_number(2, 11)
        self.assertEqual(2, second_episode['season'])
        self.assertEqual(11, second_episode['number'])

    def test_invalid_date_episode_by_date(self):
        """bad date for episode_by_date() should raise ValueError"""
        show = Show(show_id=1)
        self.assertRaises(ValueError, lambda: show.episodes_by_date('12'))
        self.assertRaises(ValueError, lambda: show.episodes_by_date(''))
        self.assertRaises(ValueError, lambda: show.episodes_by_date(1))
        self.assertRaises(ValueError, lambda: show.episodes_by_date('2002-02-21'))

    def test_valid_date_episode_by_date(self):
        """valid date for episode_by_date() should return list of episodes"""
        show = Show(show_id=1)
        episodes = show.episodes_by_date('2013-07-01')
        for episode in episodes:
            self.assertEqual('2013-07-01', episode['airdate'])

    def test_cast(self):
        """cast property should return a list of cast members"""
        show = Show(show_id=1)
        self.assertTrue('/1/cast' in show.cast_url)
        self.assertFalse(show.cast_list)
        cast_list = show.cast
        self.assertTrue(show.cast_list)
        self.assertTrue(cast_list)

        for cast in cast_list:
            self.assertTrue(hasattr(cast, 'character'))
            self.assertIsInstance(obj=cast, cls=Cast)

    def test_cast_with_embed_url(self):
        """cast property should return list of cast members as Cast objects when created with embed_url"""
        show = Show(show_id=1, embed_url="?embed=cast")
        for cast in show.cast:
            self.assertIsInstance(cast, Cast)

    def test_crew(self):
        """crew property should return list of crew members"""
        show = Show(show_id=1)
        self.assertTrue('/1/crew' in show.crew_url)
        self.assertFalse(show.crew_list)
        crew = show.crew
        self.assertTrue(show.crew_list)
        self.assertTrue(crew)
        for person in crew:
            self.assertTrue(hasattr(person, 'crew_type'))
            self.assertIsInstance(person, Crew)

    def test_crew_with_embed_url(self):
        """crew property should return list of crew members as Crew objects when created with embed_url"""
        show = Show(show_id=1, embed_url='?embed=crew')
        for crew in show.crew:
            self.assertIsInstance(crew, Crew)

    def test_get_episode_method_with_valid_int(self):
        """get_episode(-1) should return the last Episode object in the episode list"""
        show = Show(show_id=1)
        self.assertEqual(show.episodes[-1], show.get_episode(-1))

    def test_get_episode_method_with_bad_input(self):
        """get_episode("taco") should return None"""
        show = Show(show_id=1)
        self.assertIsNone(show.get_episode('taco'))

    def test_get_episode_method_with_out_of_range_input(self):
        """get_episode(100000) should return None assuming a show doesn't have 100,000 episodes"""
        show = Show(show_id=1)
        self.assertIsNone(show.get_episode(100000))

    def test_next_episode(self):
        """next_episode property should return Episode object or None"""
        schedule = Schedule()
        show = schedule.episodes[0].show
        self.assertTrue(show.next_episode)


class TestSeason(unittest.TestCase):
    """Tests Season class in tvmaze.py"""

    def test_create_season_with_bad_season(self):
        """Tests init method in Season class"""
        self.assertRaises(ValueError, lambda: Season(season_id='t'))

    def test_season_with_good_season(self):
        """Tests init method in Season class with good season_id"""
        season = Season(season_id=1)
        self.assertTrue(season)

    def test_season_with_episodes(self):
        """Tests init method in Season class with_episodes=True"""
        season = Season(season_id=1, with_episodes=True)
        for episode in season.episodes:
            self.assertEqual(1, episode.season)

    def test_episodes(self):
        """Test episodes property"""
        season = Season(season_id=1)

        # make sure episode list empty to ensure season.episodes makes the correct api call
        self.assertFalse(season.episode_list)

        # api call made
        self.assertTrue(season.episodes)

        # make sure episode_list is correctly populated after api call
        self.assertTrue(season.episode_list)


class TestCast(unittest.TestCase):
    """Tests for Cast class in tvmaze.py"""

    def setUp(self):
        self.cast_list = make_request('http://api.tvmaze.com/shows/1/cast')

    def test_create_cast_class(self):
        """Tests init method of cast class"""
        cast_member = Cast(**self.cast_list[0])
        self.assertTrue(cast_member.character)
        self.assertTrue(cast_member.name)
        self.assertTrue(cast_member.url)
        self.assertTrue(cast_member.image)
        self.assertNotEqual('no name', cast_member.name)
        self.assertNotEqual('no url entered', cast_member.url)
        self.assertNotEqual({"medium": "no link", "original": "no link"}, cast_member.image)


class TestCrew(unittest.TestCase):
    """Tests for Crew class in tvmaze.py"""

    def setUp(self):
        self.crew_list = make_request('http://api.tvmaze.com/shows/1/crew')

    def test_create_crew_class(self):
        """Tests init method of crew class"""
        crew_member = Crew(**self.crew_list[0])
        self.assertTrue(crew_member.crew_type)
        self.assertTrue(crew_member.name)
        self.assertTrue(crew_member.url)
        self.assertTrue(crew_member.image)
        self.assertNotEqual('no name', crew_member.name)
        self.assertNotEqual('no url entered', crew_member.url)
        self.assertNotEqual({"medium": "no link", "original": "no link"}, crew_member.image)


class TestSearch(unittest.TestCase):

    def test_search_shows(self):
        search = Search(query='under the dome')
        self.assertTrue(search.results)
        self.assertEqual(search.results[0].name, Show(show_id=1).name)
        for result in search.results:
            self.assertIsInstance(result, Show)

    def test_search_single_search(self):
        """single search only returns a single result instead of a list"""
        search = Search(variant='singlesearch', query='under the doem')
        self.assertTrue(search.results)
        self.assertEqual(search.results.name, Show(show_id=1).name)
        self.assertIsInstance(search.results, Show)

    def test_search_lookup(self):
        """look up search only returns single result"""
        search = Search(variant='lookup', lookup='thetvdb=81189')
        # look up search is of breaking bad we will get breaking bad by name and compare with thetvdb lookup
        second_search = Search(variant='singlesearch', query='breaking bad')
        self.assertTrue(search.results)
        self.assertEqual(search.results.name, second_search.results.name)
        self.assertIsInstance(search.results, Show)

    def test_search_people(self):
        search = Search(variant='people', query='lauren')
        api_results = [Person(**item) for item in make_request(search.url)]
        for person1, person2 in zip(search.results, api_results):
            self.assertEqual(person1.name, person2.name)
            self.assertEqual(person1.id, person2.id)
            self.assertIsInstance(person1, Person)


pytv/tvmaze_utility.py
"""Utility functions for tvmaze"""

import requests
import os

dir_path = os.path.dirname(os.path.realpath(__file__))


class ApiError(Exception):
    """Error class for all api calls"""


def make_request(url):
    """Return JSON data from request or error

    :param url: Url to make get request
    :return: Return JSON data from request or error
    """
    r = requests.get(url)
    data = r.json()
    if r.status_code > 400:
        raise ApiError({"message": data['message'] if 'message' in data else 'Not found', "statusCode": r.status_code})
    return data


def get_list(obj, lst, url, make_class=None):
    """Helper function to get list of data"""
    attr = getattr(obj, lst)
    if attr:
        return attr
    try:
        if make_class:
            new_list = [make_class(**item) for item in make_request(url)]
        else:
            new_list = make_request(url)
    except ApiError as e:
        raise e
    else:
        setattr(obj, lst, new_list)
        return new_list


def json_to_init(url, item_name, file=None):
    item = requests.get(url).json()
    with open('{}{}empty_file.py'.format(dir_path, os.path.sep), 'w') as file:
        for key in item.keys():
            file.write('self.{key_name} = {item_key}["{key_name}"] \n'.format(key_name=key, item_key=item_name))


pytv/tvmaze.py
import requests
import time

from .apiurls import (
    SCHEDULE,
    SHOW,
    SEASON,
    EPISODE,
    SHOWS,
    UPDATES,
    PEOPLE_SEARCH,
    SINGLE_SEARCH,
    SHOW_SEARCH,
    SHOW_LOOKUP
)
from .tvmaze_utility import make_request, ApiError, get_list


def get_updates():
    return make_request(UPDATES)


class Schedule:
    """Schedule class for tvmaze schedule endpoint
    """

    def __init__(self, date=None, country_code='US'):
        """init method for Schedule

        :param date: Optional ISO 8601 formatted date defaults to current day. Example 2017-8-14
        :param country_code: Optional ISO 3166-1 two digit country code defaults to US.
        """
        self.date = date or time.strftime('%Y-%m-%d')
        self.country_code = country_code
        self.url = '{}?country={}&date={}'.format(SCHEDULE, self.country_code, self.date)
        self.episode_list = []

    @property
    def episodes(self):
        """Return list of episodes"""
        try:
            episode_list = get_list(self, 'episode_list', self.url, Episode)
        except ApiError as e:
            raise e
        else:
            return sorted(
                episode_list,
                key=lambda episode: int(episode.show.weight),
                reverse=True
            )

    def exclude_networks(self, networks=None):
        """Exclude certain networks from the schedule

        :param list networks: List of networks to exclude from schedule
        :return: List of episode objects from schedule without networks included
        """
        if not networks:
            raise ValueError("must include at least one network to exclude")
        return [episode for episode in self.episodes if episode.show.network["name"] not in networks]

    def include_networks(self, networks=None):
        """Returns list of episodes from networks included"""
        if not networks:
            raise ValueError("must include at least one network to include")
        return [episode for episode in self.episodes if episode.show.network["name"] in networks]


class Shows:
    def __init__(self, page=0):
        if page < 0:
            page = 0
        url = self.get_url(page)
        self.shows = [Show(**show) for show in make_request(url)]
        self.page = page

    def get_url(self, page=0):
        return SHOWS if not page else '{}?page={}'.format(SHOWS, page)

    def change_page(self, change):
        if self.page == change:
            return self.shows
        new_page = self.page + change
        try:
            self.shows = [Show(**show) for show in make_request(self.get_url(new_page))]
            self.page = new_page
        except ApiError as e:
            raise e
        else:
            return self.shows

    def next_page(self):
        self.change_page(1)
        if not self.shows:
            raise ValueError('No more shows')

    def previous_page(self):
        self.change_page(-1)


class Show:
    """Show class for tvmaze api"""

    def __init__(self, show_id=None, embed_url=None, **kwargs):
        self.cast_list = []
        self.crew_list = []
        self.season_list = []
        self.episode_list = []
        self.special_list = []
        self.next = None

        if show_id:
            self.show_id = show_id

            self.api_url = '{}{}{}'.format(SHOW, show_id, embed_url if embed_url else '')
            try:
                show = make_request(self.api_url)
            except ApiError as e:
                raise e
            else:
                self.schedule = show["schedule"]
                self.updated = show["updated"]
                self.status = show["status"]
                self.webChannel = show["webChannel"]
                self.premiered = show["premiered"]
                self.externals = show["externals"]
                self.summary = show["summary"]
                self.image = show["image"]
                self.type = show["type"]
                self.genres = show["genres"]
                self.runtime = show["runtime"]
                self._links = show["_links"]
                self.rating = show["rating"]
                self.officialSite = show["officialSite"]
                self.weight = show["weight"]
                self.language = show["language"]
                self.id = show["id"]
                self.url = show["url"]
                self.network = show["network"]
                self.name = show["name"]

                #: optional _embedded for embed_url values
                if embed_url:
                    self._embedded = show["_embedded"]
                    if 'episodes' in self._embedded:
                        self.episode_list = [Episode(**episode) for episode in self._embedded["episodes"]]
                    if 'cast' in self._embedded:
                        self.cast_list = [Cast(**cast) for cast in self._embedded["cast"]]
                    if 'crew' in self._embedded:
                        self.crew_list = [Crew(**crew) for crew in self._embedded["crew"]]
                    if 'seasons' in self._embedded:
                        self.season_list = [Season(**season) for season in self._embedded["seasons"]]
        else:
            self.show_id = kwargs.get('id')
            for key, val in kwargs.items():
                setattr(self, key, val)

        self.base_url = '{}{}'.format(SHOW, self.show_id)
        self.crew_url = '{}/crew'.format(self.base_url)
        self.cast_url = '{}/cast'.format(self.base_url)
        self.episodes_url = '{}/episodes'.format(self.base_url)
        self.season_url = '{}/seasons'.format(self.base_url)

    def __str__(self):
        return "{}".format(self.name)

    @property
    def episodes(self):
        """Return list of episodes"""
        try:
            return get_list(self, 'episode_list', self.episodes_url, Episode)
        except ApiError as e:
            raise e

    def get_episode(self, index):
        """Return episode in episode_list based on index

        :param int index: Index of episode list
        :return: Episode object
        """
        try:
            return self.episodes[int(index)]
        except (ValueError, IndexError):
            return None

    @property
    def next_episode(self):
        """Return episode that will air next"""
        if self.next:
            return self.next
        try:
            episode_url = self._links['nextepisode']['href']
        except KeyError:
            return None
        else:
            self.next = Episode(create_url=episode_url)
            return self.next

    @property
    def specials(self):
        """Return list of episodes with specials if show has specials"""
        try:
            return [
                episode for episode in
                get_list(self, 'special_list', '{}?specials=1'.format(self.episodes_url), Episode)
                if not episode.number
            ]
        except ApiError as e:
            raise e

    def episode_by_number(self, season, episode):
        """Return episode object

        :param int season: Season number of show. Example 1
        :param int episode: Episode number of show. Example 2
        :return: Return episode object
        """
        episode = requests.get(
            '{}{}/episodebynumber?season={}&number={}'.format(SHOW, self.show_id, season, episode)
        )
        if episode.status_code == 404:
            raise ValueError("Episode does not exist")
        return episode.json()

    def episodes_by_date(self, date):
        """Return all episodes that have aired on a specific date

        :param date: an ISO 8601 formatted date. Example 2016-08-24
        :return: List of dicts
        """
        episodes = requests.get(
            '{}{}/episodesbydate?date={}'.format(SHOW, self.show_id, date)
        )
        if episodes.status_code >= 400:
            raise ValueError('{}'.format(episodes.json()['message']))
        return episodes.json()

    @property
    def seasons(self):
        """Return list of seasons related to the show"""
        try:
            return get_list(self, 'season_list', self.season_url, Season)
        except ApiError as e:
            raise e

    @property
    def cast(self):
        """Return list of cast related to the show"""
        try:
            return get_list(self, 'cast_list', self.cast_url, Cast)
        except ApiError as e:
            raise e

    @property
    def crew(self):
        """Return list of crew related to the show"""
        try:
            return get_list(self, 'crew_list', self.crew_url, Crew)
        except ApiError as e:
            raise e


class Season:
    def __init__(self, season_id=None, with_episodes=False, **kwargs):
        """Create new season with season_id or kwargs

        :param int season_id: Id of season not required if creating season from kwargs
        :param bool with_episodes: Include episodes with season
        """
        self.episode_list = []
        if season_id:
            self.season_id = season_id
            self.api_url = "{}{}{}".format(SEASON, self.season_id, "?embed=episodes" if with_episodes else '')
            response = requests.get(self.api_url)
            if response.status_code > 400:
                raise ValueError('{}'.format(response.json()['message']))
            season = response.json()
            if with_episodes:
                self.episode_list = [Episode(**episode) for episode in season['_embedded']['episodes']]
            self.name = season["name"]
            self._links = season["_links"]
            self.url = season["url"]
            self.premiereDate = season["premiereDate"]
            self.id = season["id"]
            self.episodeOrder = season["episodeOrder"]
            self.number = season["number"]
            self.endDate = season["endDate"]
            self.webChannel = season["webChannel"]
            self.summary = season["summary"]
            self.network = season["network"]
            self.image = season["image"]
        else:
            self.season_id = kwargs.get('id')
            for key, val in kwargs.items():
                setattr(self, key, val)
        self.base_url = "{}{}".format(SEASON, self.season_id)

    @property
    def episodes(self):
        if self.episode_list:
            return self.episode_list
        r = requests.get('{}/episodes'.format(self.base_url))
        if r.status_code >= 400:
            return []
        self.episode_list = [Episode(**episode) for episode in r.json()]
        return self.episode_list


class Episode:
    """Episode class for tvmaze api"""
    def __init__(self, create_url=None, **kwargs):
        """Create episode with episode url or a dict of episode info"""
        if create_url:
            try:
                episode = make_request(create_url)
            except ApiError as e:
                raise e
            else:
                self.airdate = episode["airdate"]
                self.airstamp = episode["airstamp"]
                self.number = episode["number"]
                self.id = episode["id"]
                self.season = episode["season"]
                self.runtime = episode["runtime"]
                self.name = episode["name"]
                self.image = episode["image"]
                self.summary = episode["summary"]
                self.airtime = episode["airtime"]
                self.url = episode["url"]
                self._links = episode["_links"]
        for key, val in kwargs.items():
            setattr(self, key, val)
        try:
            self.show = Show(**kwargs['show'])
        except KeyError:
            self.show = None


class Person:
    """Person class for tvmaze api"""

    def __init__(self, person, **kwargs):
        self.id = person.get('id', 1)
        self.url = person.get('url', 'no url entered')
        self.name = person.get('name', 'no name')
        self.image = person.get('image', {"medium": "no link", "original": "no link"})
        for key, val in kwargs.items():
            setattr(self, key, val)

    def __str__(self):
        return self.name


class Crew(Person):
    def __init__(self, type, **kwargs):
        super().__init__(**kwargs)
        self.crew_type = type


class Cast(Person):
    def __init__(self, character, **kwargs):
        super().__init__(**kwargs)
        self.character = character


class Search:
    def __init__(self, query=None, variant='shows', **kwargs):
        """

        variant options:

        - shows(default)
        - singlesearch
        - lookup
        - people

        :param str variant: One of the api search options on tv maze.
        :param str query: Search query.
        """
        self.variant = variant
        self.results_list = None
        self.kls = Show
        if variant == 'shows':
            self.url = '{}?q={}'.format(SHOW_SEARCH, query)
        elif variant == 'singlesearch':
            url = '{}?q={}'.format(SINGLE_SEARCH, query)
            embed_string = kwargs.get('embed_string')
            url += '&{}'.format(embed_string)
            self.url = url
        elif variant == 'lookup':
            lookup = kwargs.get('lookup')
            if not lookup:
                raise ValueError("""
                lookup string i.e tvrage=1242 required when search is variant lookup.
                See documentation for other lookup strings
                """)
            self.url = '{}?{}'.format(SHOW_LOOKUP, lookup)
        elif variant == 'people':
            self.url = '{}?q={}'.format(PEOPLE_SEARCH, query)
            self.kls = Person

    @property
    def results(self):
        if self.results_list:
            return self.results_list
        if self.variant == 'singlesearch' or self.variant == 'lookup':
            self.results_list = self.kls(**make_request(self.url))
        elif self.variant == 'people':
            self.results_list = [self.kls(**item) for item in make_request(self.url)]
        else:
            self.results_list = [self.kls(**item['show']) for item in make_request(self.url)]
        return self.results_list
